Mathlib/RingTheory/LaurentSeries.lean

```diff
@@ -852,4 +876,300 @@ instance instLaurentSeriesComplete : CompleteSpace K⸨X⸩ :=
+  set ψ := (Valued.v : RatFunc K → ℤₘ₀) with hψ
+  have := Valued.is_topological_valuation (R := (Polynomial.idealX K).adicCompletion (RatFunc K))
+  by_cases ha : a = 0
+  · rw [tendsto_def]
+    intro S hS
+    rw [ha, map_zero, WithZeroTopology.hasBasis_nhds_zero.1 S] at hS
+    obtain ⟨γ, γ_ne_zero, γ_le⟩ := hS
+    use {t | Valued.v t < γ}
+    constructor
+    · rw [ha, this]
+      use Units.mk0 γ γ_ne_zero
+      rw [Units.val_mk0]
+    · refine Set.Subset.trans (fun a _ ↦ ?_) (Set.preimage_mono γ_le)
+      rwa [Set.mem_preimage, Set.mem_Iio, hψ, ← Valued.valuedCompletion_apply a]
+  · rw [WithZeroTopology.tendsto_of_ne_zero ((Valuation.ne_zero_iff Valued.v).mpr ha), hψ,
+      Filter.eventually_comap, Filter.Eventually, Valued.mem_nhds]
+    set γ := Valued.v a / (↑(Multiplicative.ofAdd (1 : ℤ)) : ℤₘ₀) with h_aγ
+    have γ_ne_zero : γ ≠ 0 := by
+      rw [ne_eq, _root_.div_eq_zero_iff, Valuation.zero_iff]
+      simpa only [coe_ne_zero, or_false]
+    use Units.mk0 γ γ_ne_zero
+    intro y val_y b diff_b_y
+    replace val_y : Valued.v y = Valued.v a := by
+      refine Valuation.map_eq_of_sub_lt _ (val_y.trans ?_)
+      rw [Units.val_mk0, h_aγ, ← coe_unzero ((Valuation.ne_zero_iff Valued.v).mpr ha), ←
+        WithZero.coe_div, coe_lt_coe, div_lt_self_iff, ← ofAdd_zero,
+        Multiplicative.ofAdd_lt]
+      exact Int.zero_lt_one
+    rw [← Valued.extension_extends, ← val_y, ← diff_b_y]
+    congr
+
+/- The extension of the `X`-adic valuation from `RatFunc K` up to its abstract completion coincides,
+modulo the isomorphism with `K⸨X⸩`, with the `X`-adic valuation on `K⸨X⸩`. -/
+theorem valuation_compare (f : K⸨X⸩) :
+    (Valued.v : (RatFuncAdicCompl K) → ℤₘ₀)
+        (AbstractCompletion.compare (LaurentSeriesPkg K) ratfuncAdicComplPkg f) =
+      Valued.v f := by
+  rw [← valuation_LaurentSeries_equal_extension, ← compare_comp_eq_compare
+    (pkg := ratfuncAdicComplPkg) (cont_f := Valued.continuous_valuation)]
+  · rfl
+  exact (tendsto_valuation K)
+
+section PowerSeries
+
+/-- In order to compare `K⟦X⟧` with the valuation subring in the `X`-adic completion of
+`RatFunc K` we consider its alias as a subring of `K⸨X⸩`. -/
+abbrev powerSeries_as_subring : Subring K⸨X⸩ :=
+  RingHom.range (HahnSeries.ofPowerSeries ℤ K)
+
+/-- The ring `K⟦X⟧` is isomorphic to the subring `powerSeries_as_subring K` -/
+abbrev powerSeriesEquivSubring : K⟦X⟧ ≃+* powerSeries_as_subring K := by
+  rw [powerSeries_as_subring, RingHom.range_eq_map]
+  exact ((Subring.topEquiv).symm).trans (Subring.equivMapOfInjective ⊤ (ofPowerSeries ℤ K)
+    ofPowerSeries_injective)
+
+/- Through the isomorphism `LaurentSeriesRingEquiv`, power series land in the unit ball inside the
+completion of `RatFunc K`. -/
+theorem mem_integers_of_powerSeries (F : K⟦X⟧) :
+    (LaurentSeriesRingEquiv K) F ∈ (Polynomial.idealX K).adicCompletionIntegers (RatFunc K) := by
+  have : (LaurentSeriesRingEquiv K) F =
```

Comment:
```suggestion
    (LaurentSeriesRingEquiv K) F ∈ (idealX K).adicCompletionIntegers (RatFunc K) := by
```

Reply:
Mmm, something went wrong with the VS Code extension, I think I just wanted to remove the `Polynomial.`